elastic/kibana
# [Stateful sidenav] Add solutionId in base path

In this PR I've added solution base path to Kibana apps following the same mechanism that is used for the space base path.

This will allow us to know which side navigation solution to render when accessing an app (e.g the "discover" app is part of each solution IA. When refreshing the browser window we need to know which solution to render).

## Why?

Some apps (e.g. "discover") are part of multiple solutions. Having the solution as part of the base path is needed for the following requirements:

* We want to be able to refresh the browser window and land on the same solution the user was
* We want to be able to share a URL which contains the solution context
* For telemetry we want to know the session length of each app per solution

## Implementation details

* Path for an observability solution app: `http://localhost:5601/ryz/n/oblt/app/discover`
* Path in observability solution app in the "hello" space `http://localhost:5601/ryz/s/hello/n/oblt/app/discover`

The following areas of the code have been updated

### Core  `BasePath` service

The `BasePath`  service has been updated. The `set()` method allows now **partial** base paths definition to be passed for a specific Request. This allows multiple parts of Kibana to set slice of the base path.

### Space plugin

The space plugin client code has been updated to take into account that there can be multiple base paths defined for the current page and we can't anymore rely just on the server base path.

The server side has been updated to allow multiple base path to be declared for the current request.

### Navigation plugin

The server side has a HTTP pre routing interceptor that will set a partial base path for the current request. 
It also detects if the global setting is enabled or disabled. If disabled and there is a solution base path in the current request path it will remove it and redirect to the same path without the solution base path.

## How to test

* Add the following setting to your `kibana.yml` file:

```yaml
xpack.cloud.id: "ftr_fake_cloud_id:aGVsbG8uY29tOjQ0MyRFUzEyM2FiYyRrYm4xMjNhYmM="
xpack.cloud.base_url: "https://cloud.elastic.co"
xpack.cloud.deployment_url: "/deployments/deploymentId"

xpack.cloud_integrations.experiments.enabled: true
xpack.cloud_integrations.experiments.flag_overrides:
  "navigation.solutionNavEnabled": true
```

* Start Kibana and navigate to the Advanced settings > Global settings tab
* Enable the "Solution navigation" setting
* Refresh the page, you should see the solution id in the URL
* In the Global search, search for "Spaces" and navigate to the "Spaces" app
* Create a new space
* Logout
* Log back in, you will be in the space selector screen
* Select the space you created
* You will be redirected to the home page of the space
* Check that both base path are present in the URL, the space and solution base paths.
